projs/shadow/shadow-engine/core/inc/shadow/core/jobs/Job.h

```diff
@@ -0,0 +1,213 @@
+     */
+    struct PriorityQueue {
+        size_t nThreads;                            // The number of threads allocated to this priority.
+        std::vector<std::thread> threads;           // The actual threads executing code for this priority.
+        std::unique_ptr<TaskQueue[]> queues;        // A list of TaskQueues, one for every thread.
+        std::atomic<uint32_t> nextQueue { 0 };      // The next TaskQueue to process.
+        std::condition_variable wake;               // The condition that, when passed, will wake any queues depending on the current.
+        std::mutex locker;
+
+        /**
+         * Process all waiting threads.
+         * @param firstQueue the first queue to perform work on, enabling skipping queues already processed.
+         */
+        inline void Execute(size_t firstQueue) {
+            TaskGroup task;
+            for (size_t i = 0; i < nThreads; i++) {
+                TaskQueue& tasks = queues[firstQueue & nThreads];
+                while (tasks.Pop(task))
+                    task.Execute();
+                firstQueue++;
+            }
+        }
+    };
+
+    /**
+     * Initialize the Task management system.
+     * @param maxThreads the maximum number of threads that can be dedicated entirely to running Tasks.
+     */
+    void Init(uint32_t maxThreads = 0xFFFFFFFF);
+
+    /**
+     * Prepare the Task management system for engine shutdown.
+     */
+    void Destroy();
+
+    /**
+     * Fetch the number of Threads currently running, for the given priority.
+     * @param p the priority of threads to check
+     * @return The number of threads currently running a task with the given priority.
+     */
+    size_t GetThreadsOfPriority(Priority p = Priority::HIGHEST);
+
+    /**
+     * Run the given task off-thread, with the given priority.
+     * The function will run, and then return. The task cannot be restarted, unless you call Run again.
+     * @param context a task context, with priority
+     * @param task the function to run off-thread
+     */
+    void Run(ExecutionContext& context, const std::function<void(TaskArguments)>& task);
+
+    /**
+     * Dispatch a set of jobs, running the same code with different arguments.
+     * The arguments that change are the task index and group index.
+     * Usually use this where iterating over a static list and performing an expensive task on every object within.
+     * @param context a task context, with priority
+     * @param jobs the number of jobs to create, total. Usually, max index of the list being processed
+     * @param groups the number of jobs to run per thread. Usually, jobs / max_threads if jobs > max_threads, or 5 otherwise.
+     * @param task the function to run per index
+     * @param sharedMem the size of a shared memory buffer to allocate for every group.
+     */
+    void Dispatch(ExecutionContext& context, size_t jobs, size_t groups, const std::function<void(TaskArguments)>& task, size_t sharedMem = 0);
```

Comment:
Would rename `jobs` to `count`
Could we provide a default grouping? Seams pointless to write out the `jobs / max_threads if jobs > max_threads, or 5 otherwise` part every time...

Reply:
In most cases, it'll be one. I can move it to the end of the function as a default parameter if it'll help.

Reply:
Now that is just more confusing cause you say in most cases it will be `1` but the comment says otherwise...

Also what about the `jobs` to `count` rename?
```suggestion
    void Dispatch(ExecutionContext& context, size_t count, size_t groups, const std::function<void(TaskArguments)>& task, size_t sharedMem = 0);
```